### PR TITLE
Full node changes required for CAT / Standalone

### PR DESCRIPTION
**Pull request by Yostra**

### CODE REVIEW

File: chia/consensus/blockchain.py [hunk 1]
```diff
@@ -286,17 +288,34 @@ async def receive_block(
...
-            return ReceiveBlockResult.ADDED_AS_ORPHAN, None, None, []
+            return ReceiveBlockResult.ADDED_AS_ORPHAN, None, None, ([], {})
+
+    def get_hint_list(self, npc_result: NPCResult) -> List[Tuple[str, str]]:
+        h_list = []
+        for npc in npc_result.npc_list:
+            for opcode, conditions in npc.conditions:
+                if opcode == ConditionOpcode.CREATE_COIN:
+                    for condition in conditions:
+                        if len(condition.vars) > 2:
```
Comment (arvidn) [inline chia/consensus/blockchain.py:301]: the next version of `clvm_rs` will actually always include the 3rd field here, but it will be an empty bytes buffer for coins that didn't specify it. So you'll need to check `condition.vars[2] != b""` as well

File: chia/consensus/blockchain.py [hunk 2]
```diff
@@ -164,7 +169,9 @@ async def receive_block(
         block: FullBlock,
         pre_validation_result: Optional[PreValidationResult] = None,
         fork_point_with_peak: Optional[uint32] = None,
-    ) -> Tuple[ReceiveBlockResult, Optional[Err], Optional[uint32], List[CoinRecord]]:
+    ) -> Tuple[
+        ReceiveBlockResult, Optional[Err], Optional[uint32], Tuple[List[CoinRecord], Dict[bytes, List[CoinRecord]]]
```
Comment (arvidn) [inline chia/consensus/blockchain.py:173]: where is the call to this function where you do something with the additional return values?

File: chia/consensus/blockchain.py [hunk 3]
```diff
@@ -286,17 +288,34 @@ async def receive_block(
...
+        h_list = []
+        for npc in npc_result.npc_list:
+            for opcode, conditions in npc.conditions:
+                if opcode == ConditionOpcode.CREATE_COIN:
+                    for condition in conditions:
+                        if len(condition.vars) > 2:
+                            puzzle_hash, amount_bin = condition.vars[0], condition.vars[1]
+                            amount = int_from_bytes(amount_bin)
+                            coin_id = Coin(npc.coin_name, puzzle_hash, amount).name().hex()
+                            h_list.append((coin_id, condition.vars[2].hex()))
```
[the remark this answers is not in the document]
Comment (Yostra) [reply chia/consensus/blockchain.py]: changed

File: chia/consensus/blockchain.py (hunk 2 above)
Comment (Yostra) [reply chia/consensus/blockchain.py]: `receive_block_batch` and `respond_block` in ful_node.py call this function and then pass the results to `update_wallets`

File: chia/consensus/blockchain.py [hunk 4]
```diff
@@ -286,17 +290,37 @@ async def receive_block(
...
+            return ReceiveBlockResult.ADDED_AS_ORPHAN, None, None, ([], {})
+
+    def get_hint_list(self, npc_result: NPCResult) -> List[Tuple[bytes, bytes]]:
+        h_list = []
+        for npc in npc_result.npc_list:
+            for opcode, conditions in npc.conditions:
+                if opcode == ConditionOpcode.CREATE_COIN:
+                    for condition in conditions:
+                        if len(condition.vars) > 2 and condition.vars[2] != b"":
+                            puzzle_hash, amount_bin = condition.vars[0], condition.vars[1]
```
Comment (mariano54) [inline chia/consensus/blockchain.py:304]: @Quexington was saying that the 3rd argument should be a list, so we can add more things there later. This is also how I am setting the memo for retirement of CAT tokens: 
https://github.com/Chia-Network/chia-blockchain/pull/8600

File: chia/consensus/blockchain.py [hunk 5]
```diff
@@ -372,46 +398,63 @@ async def _reconsider_peak(
...
+                    record: Optional[CoinRecord]
+                    for record in added_rec:
+                        assert record
+                        lastest_coin_state[record.name] = record
+                    for record in removed_rec:
+                        assert record
+                        lastest_coin_state[record.name] = record
+
+                    if npc_res is not None:
+                        hint_list = self.get_hint_list(npc_res)
```
Comment (mariano54) [inline chia/consensus/blockchain.py:434]: Can you add a type annotation here? Not sure what type this list is.

File: chia/consensus/blockchain.py [hunk 6]
```diff
@@ -286,17 +290,37 @@ async def receive_block(
         if fork_height is not None:
             # new coin records added
             assert coin_record_change is not None
-            return ReceiveBlockResult.NEW_PEAK, None, fork_height, coin_record_change
+            return ReceiveBlockResult.NEW_PEAK, None, fork_height, (coin_record_change, hint_changes)
         else:
-            return ReceiveBlockResult.ADDED_AS_ORPHAN, None, None, []
+            return ReceiveBlockResult.ADDED_AS_ORPHAN, None, None, ([], {})
+
+    def get_hint_list(self, npc_result: NPCResult) -> List[Tuple[bytes, bytes]]:
```
Comment (mariano54) [inline chia/consensus/blockchain.py:297]: coin_id is type bytes32 instead of bytes

File: chia/consensus/blockchain.py [hunk 7]
```diff
@@ -372,46 +398,63 @@ async def _reconsider_peak(
...
+                        lastest_coin_state[record.name] = record
+                    for record in removed_rec:
+                        assert record
+                        lastest_coin_state[record.name] = record
+
+                    if npc_res is not None:
+                        hint_list = self.get_hint_list(npc_res)
+                        await self.hint_store.add_hints(hint_list)
+                        # There can be multiple coins for the same hint
+                        for coin_name, hint in hint_list:
```
Comment (mariano54) [inline chia/consensus/blockchain.py:437]: coin_id instead of coin_name

File: chia/consensus/blockchain.py [hunk 8]
```diff
@@ -372,46 +398,63 @@ async def _reconsider_peak(
...
+                            key = hint
```
Comment (mariano54) [inline chia/consensus/blockchain.py:438]: Can we remove the "key" variable?

File: chia/full_node/hint_store.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    async def get_hints(self, hint: bytes) -> List[bytes32]:
+        cursor = await self.coin_record_db.execute("SELECT * from hints WHERE hint=?", (hint,))
+        rows = await cursor.fetchall()
+        await cursor.close()
+        coin_ids = []
+        for row in rows:
+            coin_ids.append(row[0])
+        return coin_ids
+
+    async def add_hints(self, coin_hint_list: List[Tuple[bytes, bytes]]) -> None:
```
Comment (mariano54) [inline chia/full_node/hint_store.py:34]: coin_id is bytes32.

File: chia/consensus/blockchain.py [hunk 10]
```diff
@@ -286,17 +290,37 @@ async def receive_block(
...
+        h_list = []
+        for npc in npc_result.npc_list:
+            for opcode, conditions in npc.conditions:
+                if opcode == ConditionOpcode.CREATE_COIN:
+                    for condition in conditions:
+                        if len(condition.vars) > 2 and condition.vars[2] != b"":
+                            puzzle_hash, amount_bin = condition.vars[0], condition.vars[1]
+                            amount = int_from_bytes(amount_bin)
+                            coin_id = Coin(npc.coin_name, puzzle_hash, amount).name()
+                            h_list.append((coin_id, condition.vars[2]))
```
Comment (mariano54) [inline chia/consensus/blockchain.py:307]: We should only add the memos which are less than 32 bytes right? Otherwise this is a DOS vector.
I suggest the following naming: The list of bytes in the 3rd argument should be called the memos, and the memos are stored as hints only if they are <= 32 bytes.

Comment (Yostra) [reply chia/consensus/blockchain.py:307]: clvm_rs doesn't return anything longer than 32bytes

File: chia/consensus/blockchain.py (hunk 8 above)
Comment (Yostra) [reply chia/consensus/blockchain.py:438]: avoids hashing same thing multiple times

File: chia/consensus/blockchain.py (hunk 4 above)
Comment (Yostra) [reply chia/consensus/blockchain.py:304]: I think this needs to be changed in clvm_rs first, was there change there?

Comment (Yostra) [reply chia/consensus/blockchain.py:304]: Just saw Arvids pr is open, I'll have to wait for that to get in & clvm_rs pypy release is available.

File: chia/full_node/full_node_api.py [hunk 11]
```diff
@@ -1391,3 +1407,39 @@ async def request_children(self, request: wallet_protocol.RequestChildren) -> Op
         response = wallet_protocol.RespondChildren(states)
         msg = make_msg(ProtocolMessageTypes.respond_children, response)
         return msg
+
+    @api_request
+    async def request_ses_hashes(self, request: wallet_protocol.RequestSESInfo):
```
Comment (arvidn) [inline chia/full_node/full_node_api.py:1412]: we can receive this request from an untrusted peer, right? as someone not very familiar with `get_ses_heights()`, it's hard to say whether this is safe, or whether an untrusted peer can make us do an unreasonable amount of work.

`get_ses_heights()` in particular, looks scary, as it doesn't take any parameters.

File: chia/full_node/hint_store.py [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+    coin_record_db: aiosqlite.Connection
+    db_wrapper: DBWrapper
+
+    @classmethod
+    async def create(cls, db_wrapper: DBWrapper):
+        self = cls()
+        self.db_wrapper = db_wrapper
+        self.coin_record_db = db_wrapper.db
+        # the coin_id is unique, same hint can be used for multiple coins
+        await self.coin_record_db.execute(("CREATE TABLE IF NOT EXISTS hints(coin_id blob PRIMARY KEY,  hint blob)"))
```
Comment (arvidn) [inline chia/full_node/hint_store.py:20]: if we ever want to support multiple hints per coin, changing this schema may be the biggest hurdle

File: chia/full_node/hint_store.py [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+        await self.coin_record_db.execute("CREATE INDEX IF NOT EXISTS hint_index on hints(hint)")
+        await self.coin_record_db.commit()
+        return self
+
+    async def get_hints(self, hint: bytes) -> List[bytes32]:
```
Comment (arvidn) [inline chia/full_node/hint_store.py:25]: this function is called `get_hints()` but returns coin-IDs. It seems better names would be: `lookup_hints()` or `get_coin_ids_from_hints()`

File: chia/protocols/protocol_message_types.py [hunk 14]
```diff
@@ -95,3 +95,5 @@ class ProtocolMessageTypes(Enum):
     respond_to_coin_update = 73
     request_children = 74
     respond_children = 75
+    request_ses_hashes = 76
```
Comment (arvidn) [inline chia/protocols/protocol_message_types.py:98]: what's `ses` short for?

File: chia/protocols/shared_protocol.py [hunk 15]
```diff
@@ -5,7 +5,7 @@
 from chia.util.ints import uint8, uint16
 from chia.util.streamable import Streamable, streamable
 
-protocol_version = "0.0.32"
+protocol_version = "0.0.33"
```
Comment (arvidn) [inline chia/protocols/shared_protocol.py:8]: do we record which messages are available in what version of the protocol anywhere?

File: tests/wallet/simple_sync/test_simple_sync_protocol.py [hunk 16]
```diff
@@ -452,3 +458,72 @@ async def test_subscribe_for_coin_id_reorg(self, wallet_node_simulator):
...
+                break
+
+        assert notified_state is not None
+        assert notified_state.items[0].coin == Coin(coin_spent.name(), puzzle_hash, amount)
+
+        msg = wallet_protocol.RegisterForPhUpdates([hint], 0)
+        msg_response = await full_node_api.register_interest_in_puzzle_hash(msg, fake_wallet_peer)
+        assert msg_response.type == ProtocolMessageTypes.respond_to_ph_update.value
+        data_response: RespondToPhUpdates = RespondToCoinUpdates.from_bytes(msg_response.data)
+        assert len(data_response.coin_states) == 1
```
Comment (arvidn) [inline tests/wallet/simple_sync/test_simple_sync_protocol.py:529]: it seems like you should check the actual value of the `coin_states` too, not just that you got *something*

**Review by arvidn — COMMENTED** [review]
perhaps the tuple of coin updates and new hints should be put in a new class holding state updates/subscription updates. Such a class could also do the deduplication and aggregation that you're currently doing "ad hoc" (or, sprinkled into the existing code)

File: chia/consensus/blockchain.py [hunk 17]
```diff
@@ -372,46 +398,63 @@ async def _reconsider_peak(
             records_to_add = []
             for fetched_full_block, fetched_block_record in reversed(blocks_to_add):
                 records_to_add.append(fetched_block_record)
-                if fetched_block_record.is_transaction_block:
+                if fetched_full_block.is_transaction_block():
```
[the remark this answers is not in the document]
Comment (Yostra) [reply chia/consensus/blockchain.py:401]: yea, it just saved on indentation bellow

File: chia/full_node/hint_store.py (hunk 12 above)
Comment (Yostra) [reply chia/full_node/hint_store.py:20]: good catch, I'll add another auto-increment id column there so that coin id is not unique

File: chia/protocols/protocol_message_types.py (hunk 14 above)
Comment (Yostra) [reply chia/protocols/protocol_message_types.py:98]: sub epoch summary

File: tests/wallet/simple_sync/test_simple_sync_protocol.py (hunk 16 above)
Comment (Yostra) [reply tests/wallet/simple_sync/test_simple_sync_protocol.py:529]: added

File: chia/protocols/shared_protocol.py (hunk 15 above)
Comment (Yostra) [reply chia/protocols/shared_protocol.py:8]: No yet

File: chia/full_node/hint_store.py (hunk 12 above)
Comment (Yostra) [reply chia/full_node/hint_store.py]: added

File: chia/consensus/blockchain.py [hunk 18]
```diff
@@ -164,7 +168,12 @@ async def receive_block(
         block: FullBlock,
         pre_validation_result: Optional[PreValidationResult] = None,
         fork_point_with_peak: Optional[uint32] = None,
-    ) -> Tuple[ReceiveBlockResult, Optional[Err], Optional[uint32], List[CoinRecord]]:
+    ) -> Tuple[
+        ReceiveBlockResult,
+        Optional[Err],
+        Optional[uint32],
+        Tuple[List[CoinRecord], Dict[bytes, Dict[bytes32, CoinRecord]]],
```
[the remark this answers is not in the document]
Comment (Yostra) [reply chia/consensus/blockchain.py:175]: I took a shortcut here, receive_block get's called on 100s of places in tests and if another arg was added in the call I would have to refactor all of those.

[the remark this answers is not in the document]
Comment (arvidn) [reply chia/consensus/blockchain.py:175]: so maybe a dedicated type for updates would make sense then. then new updates could be added without affecting this signature

**Comment by arvidn** [comment]
I retriggered the failing test. Is there a risk this patch slowed down the execution to cause that timing constraint to be exceeded?
If so, we could bump it further under the understanding that we're increasing the tech debt

File: chia/consensus/blockchain.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (mariano54) [reply chia/consensus/blockchain.py:175]: I'm not sure how this is passing the tests, given that the response is checked in a lot of tests.
I agree with arvid, this response is getting quite complex. Perhaps we can put all of these things into one struct (including the fork_height). 

We can leave this for the future, but this is a pretty important method, so it would be good to keep it simple

**Comment by arvidn** [comment]
Reviewing this, I haven't found any obvious errors, but I'm also not really convinced that all new edge cases introduced by this are covered by tests.

File: chia/consensus/blockchain.py [hunk 19]
```diff
@@ -372,46 +398,63 @@ async def _reconsider_peak(
             records_to_add = []
             for fetched_full_block, fetched_block_record in reversed(blocks_to_add):
                 records_to_add.append(fetched_block_record)
-                if fetched_block_record.is_transaction_block:
+                if fetched_full_block.is_transaction_block():
                     if fetched_block_record.header_hash == block_record.header_hash:
-                        tx_removals, tx_additions = await self.get_tx_removals_and_additions(
+                        tx_removals, tx_additions, npc_res = await self.get_tx_removals_and_additions(
                             fetched_full_block, npc_result
                         )
```
[the remark this answers is not in the document]
Comment (mariano54) [reply chia/consensus/blockchain.py:403]: We don't run it again. If the npc_result already exists (it's Optional, and not present when syncing) then we will not run the generator. So we are guaranteed to only run it once.

File: chia/full_node/full_node_api.py (hunk 11 above)
Comment (mariano54) [reply chia/full_node/full_node_api.py:1412]: Blockchain.get_ses_heights() is very fast since it stores these things in memory (the height of each sub-epoch)

Comment (mariano54) [reply chia/full_node/full_node_api.py:1412]: Everything here seems safe and very fast. However, we should have a rate limit like other messages.

**Review by mariano54 — APPROVED** [review]
LGTM after addressing arvid's feedback (and maybe adding a rate limit)
